standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -3916,56 +3879,59 @@ public ShowCompactResponse showCompact(ShowCompactRequest rqst) throws MetaExcep
       try (Connection dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
         PreparedStatement stmt = sqlGenerator.prepareStmtWithParameters(dbConn, query.toString(),
           getShowCompactionQueryParamList(rqst))) {
-          LOG.debug("Going to execute query <" + query + ">");
-          try (ResultSet rs = stmt.executeQuery()) {
-            while (rs.next()) {
-              ShowCompactResponseElement e = new ShowCompactResponseElement();
-              e.setDbname(rs.getString(1));
-              e.setTablename(rs.getString(2));
-              e.setPartitionname(rs.getString(3));
-              e.setState(compactorStateToResponse(rs.getString(4).charAt(0)));
-              try {
-                e.setType(dbCompactionType2ThriftType(rs.getString(5).charAt(0)));
-              } catch (MetaException ex) {
-                //do nothing to handle RU/D if we add another status
-              }
-              e.setWorkerid(rs.getString(6));
-              long start = rs.getLong(7);
-              if (!rs.wasNull()) {
-                e.setStart(start);
-              }
-              long endTime = rs.getLong(8);
-              if (endTime != -1) {
-                e.setEndTime(endTime);
-              }
-              e.setRunAs(rs.getString(9));
-              e.setHadoopJobId(rs.getString(10));
-              e.setId(rs.getLong(11));
-              e.setErrorMessage(rs.getString(12));
-              long enqueueTime = rs.getLong(13);
-              if (!rs.wasNull()) {
-                e.setEnqueueTime(enqueueTime);
-              }
-              e.setWorkerVersion(rs.getString(14));
-              e.setInitiatorId(rs.getString(15));
-              e.setInitiatorVersion(rs.getString(16));
-              long cleanerStart = rs.getLong(17);
-              if (!rs.wasNull() && (cleanerStart != -1)) {
-                e.setCleanerStart(cleanerStart);
-              }
-              String poolName = rs.getString(18);
-              if (isBlank(poolName)) {
-                e.setPoolName(DEFAULT_POOL_NAME);
-              } else {
-                e.setPoolName(poolName);
-              }
-              e.setTxnId(rs.getLong(19));
-              e.setNextTxnId(rs.getLong(20));
-              e.setCommitTime(rs.getLong(21));
-              e.setHightestTxnId(rs.getLong(22));
-              response.addToCompacts(e);
+        if (rqst.isSetId()) {
+          stmt.setLong(getShowCompactionQueryParamList(rqst).size() + 1, rqst.getId());
```

Comment:
Maybe introducing a QueryBuilder, like for QueryBasedCompaction? All the showCompact related query parts, `getShowCompactionQueryParamList(rqst)` and `getShowCompactionFilterClause(rqst)`, could be extracted from TxnHandler. However, not sure if it worth it. At least the result of `getShowCompactionFilterClause(rqst)` should be stored into a variable to avoid calling it twice.

Reply:
fixed